oeps/oep-0036-microservice-routing-layer.rst

```diff
@@ -0,0 +1,89 @@
+===================================
+OEP-36: Microservice Routing Layer
+===================================
+
+.. list-table::
+
+  * - OEP
+    - :doc:`oep-0036-microservice-routing-layer`
+  * - Title
+    - Microservice Routing Layer
+  * - Last-Modified
+    - 2019-04-25
+  * - Author
+    - Fred Smith <derf@edx.org>
+  * - Arbiter
+    - David Joy <djoy@edx.org>
+  * - Status
+    - Draft
+  * - Type
+    - Architecture
+  * - Created
+    - 2019-04-25
+
+Abstract
+========
+
+Proposes a key openedx component that will connect users to applications using a layer 7 (application layer/URL Based) routing framework. 
+
+Motivation
+==========
+
+Microservices are currently implemented using layer 4 (DNS) based routing. Each microservice occupies a unique hostname, with dedicated ELB or CDN infrastructure to support routing user traffic to the service. This pattern is simple to monitor, deploy, and manage, but slow and expensive to build, and regularly causes problems with cross service user information sharing. In Addition, modifying or splitting a monolithic service into microservices results in confusing URL changes and costly migrations.
+
+Specification
+=============
+
+A layer 7 routing application will be chosen that meets the following requirements:
+
+* High performance connection management capabilities, including support for slow clients or backends, and wire-speed throughput.
+* SSL termination on the router.
+* Flexible configuration backends including container service discovery, flat files, and filesystem routing.
+* Path based routing, including path subsets and regular expression matching
+
+This layer 7 routing application will be deployed on a common hostname.
+
+Rationale
+=========
+
+A layer 7 routing layer would allow us to separate the user facing URL or API endpoint from the backend implementation. It would allow microservices to exist that only manage a small number of endpoints, without requring developers to build out load balancing or CDN infrastructure. 
+
+We are currently using a layer 7 routing framework for the edx.org marketing site, allowing them to replace drupal generated pages with gatsby generated pages quickly and seamlessly.
+
+This change will allow developers, openedx operators and resellers to deploy new openedx sites quicker and easier than they currently can.  The router will provide a map for people to understand which services they need to deploy.
+
+
+
+Implementation
+==============
+
+* to be determined
```

Comment:
It seems that this OEP was written with [Traefik](https://docs.traefik.io/configuration/backends/file/), [Ambassador](https://www.getambassador.io/) and Nginx in mind. Why not mention these options here?

Reply:
Your suggested urls.json file looks a lot like I imagined this would work.  My expectation is that this will jinja2 straight into an nginx config, without ansible or any other complex tooling.  The expectation is that this will never be implemented with ansible.

I agree that using meaningful URLs is a good idea.

at some point, it will include an implementation, but I'm still gathering requirements at this stage.